internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -6,39 +6,52 @@ package githubupstreamwatcher
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
 	"fmt"
+	"net"
+	"net/http"
+	"time"
 
+	"golang.org/x/oauth2"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/errors"
 	errorsutil "k8s.io/apimachinery/pkg/util/errors"
 	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/util/cert"
 
 	"go.pinniped.dev/generated/latest/apis/supervisor/idp/v1alpha1"
 	supervisorclientset "go.pinniped.dev/generated/latest/client/supervisor/clientset/versioned"
 	idpinformers "go.pinniped.dev/generated/latest/client/supervisor/informers/externalversions/idp/v1alpha1"
 	pinnipedcontroller "go.pinniped.dev/internal/controller"
 	"go.pinniped.dev/internal/controller/conditionsutil"
+	"go.pinniped.dev/internal/controller/supervisorconfig/upstreamwatchers"
 	"go.pinniped.dev/internal/controllerlib"
+	"go.pinniped.dev/internal/endpointaddr"
 	"go.pinniped.dev/internal/federationdomain/upstreamprovider"
 	"go.pinniped.dev/internal/plog"
 	"go.pinniped.dev/internal/upstreamgithub"
 )
 
 const (
-	// Setup for the name of our controller in logs.
 	controllerName = "github-upstream-observer"
 
 	// Constants related to the client credentials Secret.
-	gitHubClientSecretType corev1.SecretType = "secrets.pinniped.dev/github-client"
-
-	//
-	// clientIDDataKey     = "clientID"
-	// clientSecretDataKey = "clientSecret"
-	//
-	// // Constants related to conditions.
-	// typeClientCredentialsValid = "ClientCredentialsValid" //nolint:gosec // this is not a credential.
+	gitHubClientSecretType               corev1.SecretType = "secrets.pinniped.dev/github-client"
+	clientIDDataKey, clientSecretDataKey string            = "clientID", "clientSecret"
+
+	HostValid                string = "HostValid"
+	TLSConfigurationValid    string = "TLSConfigurationValid"
+	OrganizationsPolicyValid string = "OrganizationsPolicyValid"
+	// ClientCredentialsObtained is different from other status conditions because it only checks that the client credentials
+	// have been obtained. The controller has no way to verify whether they are valid.
+	ClientCredentialsObtained string = "ClientCredentialsObtained" //nolint:gosec // this is not a credential
```

Comment:
Per conversation:
- GitHub OAuth Apps have an endpoint that can be called that WILL tell you if the credentials are valid (200 vs 400)
- GitHub Apps do not have the endpoint that provides this feature.

If we can find a doc link to GitHub for this we may want to note it here.  

So, perhaps we want an enhancement later for these two cases.

Reply:
https://docs.github.com/en/rest/meta/meta?apiVersion=2022-11-28#get-all-api-versions

This endpoint does not require authentication, so it is "public". However, there is severe rate limiting that would likely prevent automated programs from actually calling this endpoint. GitHub OAuth Apps can use their client id/client secret to authenticate themselves to this endpoint and receive a significantly higher rate limit. GitHub Apps cannot authenticate themselves in the same way. In fact, providing valid credentials will return 4xx, whereas providing invalid client credentials returns 200 with the lower rate limit.

There seems to be no reliable way for a GitHub App to authentication itself using only client id/secret so that it can call even public endpoints with the higher rate limit.